javalib/src/main/scala/java/util/random/RandomGenerator.scala

```diff
@@ -0,0 +1,349 @@
+    else
+      throw new IllegalArgumentException(s"Illegal bound: $bound")
+  }
+
+  def nextFloat(origin: Float, bound: Float): Float = {
+    // `origin < bound` is false if either input is NaN
+    if (origin != Float.NegativeInfinity && origin < bound && bound != Float.PositiveInfinity) {
+      val difference = bound - origin
+      val result = if (difference != Float.PositiveInfinity) {
+        // Easy case
+        origin + nextFloatBoundedInternal(difference)
+      } else {
+        /* Overflow -- we need to split the range in two.
+         * We did not manage to recover an algorithm that is compatible with the JDK.
+         * This algorithm produces values that are within 4 ULPs of the JDK.
+         */
+        val halfDiff = bound/2.0f - origin/2.0f
+        val rnd = nextFloat()
+        if (rnd < 0.5f)
+          origin + (rnd * 2.0f) * halfDiff
+        else
+          (origin + halfDiff) + ((rnd - 0.5f) * 2.0f) * halfDiff
+      }
+
+      ensureBelowBound(result, bound)
+    } else {
+      throw new IllegalArgumentException(s"Illegal bounds: [$origin, $bound)")
+    }
+  }
+
+  @inline
+  private def nextFloatBoundedInternal(bound: Float): Float =
+    nextFloat() * bound
+
+  @inline
+  private def ensureBelowBound(value: Float, bound: Float): Float = {
+    /* Based on documentation for Random.doubles to avoid issue #2144 and other
+     * possible rounding up issues:
+     * https://docs.oracle.com/javase/8/docs/api/java/util/Random.html#doubles-double-double-
+     */
+    if (value < bound) value
+    else Math.nextDown(value)
+  }
+
+  def nextDouble(): Double = {
+    // > Uses the 53 high-order bits from a call to nextLong()
+    val bits = nextLong() >>> (64 - 53)
+    bits.toDouble * (1.0 / (1L << 53)) // lossless multiplication
+  }
+
+  def nextDouble(bound: Double): Double = {
+    // false for NaN
+    if (bound > 0 && bound != Double.PositiveInfinity)
+      ensureBelowBound(nextDoubleBoundedInternal(bound), bound)
+    else
+      throw new IllegalArgumentException(s"Illegal bound: $bound")
+  }
+
+  def nextDouble(origin: Double, bound: Double): Double = {
+    // `origin < bound` is false if either input is NaN
+    if (origin != Double.NegativeInfinity && origin < bound && bound != Double.PositiveInfinity) {
+      val difference = bound - origin
+      val result = if (difference != Double.PositiveInfinity) {
+        // Easy case
+        origin + nextDoubleBoundedInternal(difference)
+      } else {
+        /* Overflow -- we need to split the range in two.
+         * We did not manage to recover an algorithm that is compatible with the JDK.
+         * This algorithm produces values that are within 4 ULPs of the JDK.
+         */
+        val halfDiff = bound/2.0 - origin/2.0
+        val rnd = nextDouble()
+        if (rnd < 0.5)
+          origin + (rnd * 2.0) * halfDiff
+        else
+          (origin + halfDiff) + ((rnd - 0.5) * 2.0) * halfDiff
```

Comment:
This seems unnecessarily complicated.  Why not just

```scala
val halfrnd = nextDouble() * halfDiff
(origin + halfrnd) + halfrnd
```

Does the compiler incorrectly optimize that to `origin + 2*halfrnd`?

Reply:
Gosh, that is indeed much simpler. I'd like to say I can't believe I didn't see that, but I think I had been fighting against these floating points for too long. 😆 

Unfortunately it's still only within 4 ULPs of the JDK generator.

Reply:
For what it's worth, this approach seems to reproduce the openjdk 21 results exactly: https://github.com/durban/choam/blob/ae2b96b444abed09728716922d92c8d168e65f77/core/shared/src/main/scala/dev/tauri/choam/random/RandomBase.scala#L174-L176 (Note, that i didn't actually look at this specific part of the openjdk source, so I don't know what they're actually doing, and I've only done limited tests with this.)

Reply:
Sure, that's another very reasonable algorithm: if the problem is too big, solve the half-problem in full, then multiply by 2.

Reply:
@durban Nice! That seems to be it!

Note about your implementation: you still need the "easy" path when the difference does not overflow, otherwise you can *underflow* `minInclusive = Double.MinPositiveValue` to `0.0`, and hence your algorithm can return a value `< minInclusive`.

Also, the dance with `doubleToLongBits` and `longBitsToDouble` is exactly what `Math.nextDown` does. ;)

Reply:
Thanks, that's a good catch with the underflow! (And yes, I should use Math.nextDown, I'm not sure why I thought it's missing in older JDKs :man_shrugging:.)

Reply:
Sure. But IMO there's still a qualitative difference between a) the distribution is not really uniform and b) it sometimes return values *outside* the request range.